contracts/Tokens/Prime/PrimeLiquidityProvider.sol

```diff
@@ -0,0 +1,209 @@
+// SPDX-License-Identifier: BSD-3-Clause
+pragma solidity 0.8.13;
+
+import { Ownable2StepUpgradeable } from "@openzeppelin/contracts-upgradeable/access/Ownable2StepUpgradeable.sol";
+import { SafeERC20Upgradeable, IERC20Upgradeable } from "@openzeppelin/contracts-upgradeable/token/ERC20/utils/SafeERC20Upgradeable.sol";
+import { AccessControlledV8 } from "@venusprotocol/governance-contracts/contracts/Governance/AccessControlledV8.sol";
+
+contract PrimeLiquidityProvider is Ownable2StepUpgradeable, AccessControlledV8 {
+    using SafeERC20Upgradeable for IERC20Upgradeable;
+
+    /// @notice The initial TOKEN index
+    uint224 public constant INITIAL_INDEX = 1e36;
+
+    /// @notice Address of the Prime contract
+    address public prime;
+
+    /// @notice The rate at which token is distributed (per block)
+    mapping(address => uint256) public tokenDistributionSpeeds;
+
+    /// @notice The rate at which token is distributed to the Prime contract
+    mapping(address => uint256) public lastAccruedBlock;
+
+    /// @notice The token accrued but not yet transferred to prime contract
+    mapping(address => uint256) public tokenAmountAccrued;
+
+    /// @dev This empty reserved space is put in place to allow future versions to add new
+    /// variables without shifting down storage in the inheritance chain.
+    uint256[46] private __gap;
+
+    /// @notice Emitted when a token distribution is initialized
+    event TokenDistributionInitialized(address indexed token);
+
+    /// @notice Emitted when a new token distribution speed is set
+    event TokenDistributionSpeedUpdated(address indexed token, uint256 newSpeed);
+
+    /// @notice Emitted when distribution state(Index and block) is updated
+    event TokensAccrued(address indexed token);
+
+    /// @notice Emitted when token is transferred to the prime contract
+    event TokenTransferredToPrime(address indexed token, uint256 amount);
+
+    /// @notice Thrown when arguments are passed are invalid
+    error InvalidArguments();
+
+    /**
+     * @param prime_ Address of the Prime contract
+     */
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor(address prime_) {
+        prime = prime_;
+
+        _disableInitializers();
+    }
+
+    /**
+     * @notice Accrue token by updating the distribution state
+     * @param token Address of the token
+     * @custom:event Emits TokensAccrued event
+     */
+    function accrueTokens(address token) public {
+        uint256 distributionSpeed = tokenDistributionSpeeds[token];
+        uint256 blockNumber = getBlockNumber();
+
+        uint256 deltaBlocks = blockNumber - lastAccruedBlock[token];
+
```

Comment:
I think we could do a "fast return" if `deltaBlocks` is 0, not even emitting the event

Reply:
Fixed